src/coprocessor/dag/expr/builtin_string.rs

```diff
@@ -650,62 +656,127 @@ impl ScalarFunc {
+            TargetLen::Invalid => Ok(None),
+            TargetLen::Len(target_len) => {
+                let r = input
+                    .iter()
+                    .chain(pad.iter().cycle())
+                    .cloned()
+                    .take(target_len)
+                    .collect::<Vec<_>>();
+                Ok(Some(Cow::Owned(r)))
+            }
         }
+    }
 
-        let r = input
-            .chars()
-            .chain(pad.chars().cycle())
-            .take(target_len)
-            .collect::<String>();
-        Ok(Some(Cow::Owned(r.into_bytes())))
+    #[inline]
+    pub fn lpad<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let input = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
+        let target_len = try_opt!(self.children[1].eval_int(ctx, row));
+        let pad = try_opt!(self.children[2].eval_string_and_decode(ctx, row));
+        let input_len = input.chars().count();
+
+        match self.valid_target_len_for_pad(target_len, 4, input_len, pad.is_empty()) {
+            TargetLen::Zero => Ok(Some(Cow::Borrowed(b""))),
+            TargetLen::Invalid => Ok(None),
+            TargetLen::Len(target_len) => {
+                let r = if let Some(remain) = target_len.checked_sub(input_len) {
+                    pad.chars()
+                        .cycle()
+                        .take(remain)
+                        .chain(input.chars())
+                        .collect::<String>()
+                } else {
+                    input.chars().take(target_len).collect::<String>()
+                };
+                Ok(Some(Cow::Owned(r.into_bytes())))
+            }
+        }
     }
 
     #[inline]
-    pub fn rpad_binary<'a, 'b: 'a>(
+    pub fn lpad_binary<'a, 'b: 'a>(
         &'b self,
         ctx: &mut EvalContext,
         row: &'a [Datum],
     ) -> Result<Option<Cow<'a, [u8]>>> {
         let input = try_opt!(self.children[0].eval_string(ctx, row));
         let target_len = try_opt!(self.children[1].eval_int(ctx, row));
         let pad = try_opt!(self.children[2].eval_string(ctx, row));
+
+        match self.valid_target_len_for_pad(target_len, 1, input.len(), pad.is_empty()) {
+            TargetLen::Zero => Ok(Some(Cow::Borrowed(b""))),
+            TargetLen::Invalid => Ok(None),
+            TargetLen::Len(target_len) => {
+                let r = if let Some(remain) = target_len.checked_sub(input.len()) {
+                    pad.iter()
+                        .cycle()
+                        .take(remain)
+                        .chain(input.iter())
+                        .cloned()
+                        .collect::<Vec<_>>()
+                } else {
+                    input[..target_len].to_vec()
+                };
+                Ok(Some(Cow::Owned(r)))
+            }
+        }
+    }
+
+    #[inline]
+    fn valid_target_len_for_pad(
```

Comment:
Pleas also add some tests for function `valid_target_len_for_pad`, thank you.

Reply:
hi @zhangjinpeng1987 this helper function is extracted  for remove duplication of the  pad-related functions， I think the tests of the for 'pad' functions is already cover it.

and I add docs to clarify what this helper function does.

Reply:
`valid_target_len_for_pad` is a individual function, i think it's better to add unit test for it.

Reply:
Comments show the function's expected behavior, tests show the exactly behavior, they are different.

Reply:
tests added, PTAL @zhangjinpeng1987